### PR TITLE
Develop upstream sync 221031

### PR DESCRIPTION
**Pull request by jayfurmanek**
https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/2849

### CODE REVIEW

**Comment by jayfurmanek** [comment]
retest Ubuntu-GPU-single please

**Comment by jayfurmanek** [comment]
retest Ubuntu-sanity please

**Comment by jayfurmanek** [comment]
retest Ubuntu-sanity please

**Comment by jayfurmanek** [comment]
retest Ubuntu-GPU-single please

**Comment by jayfurmanek** [comment]
retest Ubuntu-GPU-multi please

**Comment by jayfurmanek** [comment]
retest Ubuntu-CPU please

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc [hunk 1]
```diff
@@ -75,7 +75,7 @@ xla::Status CompileAndPrintLlvmIr(const std::string& hlo_text,
   std::string target_triple = "nvptx64-nvidia-cuda";
   std::string datalayout = "nvptx64-nvidia-cuda";
   std::string platform_name = "CUDA";
-#else
+#elif TENSORFLOW_USE_ROCM
```
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:78]: Actually, this part doesn't need #elif TENSORFLOW_USE_ROCM, and it will cause build failed. #else works here. .

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc [hunk 2]
```diff
@@ -84,8 +84,13 @@ xla::Status CompileAndPrintLlvmIr(const std::string& hlo_text,
       std::unique_ptr<llvm::Module> llvm_module,
       xla::gpu::CompileModuleToLlvmIr(
           hlo_module.get(), &llvm_context,
+#if GOOGLE_CUDA
           /*target_triple=*/xla::gpu::nvptx::TargetTriple(),
           /*data_layout=*/xla::gpu::nvptx::DataLayout(),
+#elif TENSORFLOW_USE_ROCM
+          /*target_triple=*/xla::gpu::amdgpu::TargetTriple(),
+          /*data_layout=*/xla::gpu::amdgpu::DataLayout(),
+#endif
```
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:93]: Same here, #else works here.

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc.bak [hunk 3]
```diff
@@ -0,0 +1,135 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Comment (rsanthanam-amd) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc.bak]: Is this file needed?

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc [hunk 4]
```diff
@@ -84,8 +84,13 @@ xla::Status CompileAndPrintLlvmIr(const std::string& hlo_text,
...
           /*platform_name=*/platform_name,
           stream_executor::rocm::kROCmPlatformId, gpu_device_info,
```
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:95]: This is ROCm-only, and it will cause an issue on cuda platform if we want to be generic with upstream.

Comment (jayfurmanek) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:95]: Right, hence the #ifdef there for GOOGLE_CUDA and TENSORFLOW_USE_ROCM.

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc (hunk 1 above)
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:78]: in my working branch from upstream, I use #else instread of #elif TENSORFLOW_USE_ROCM, because it actually doesn't call anything from BUILD/if_rocm_is_configured part, and it will cause the build fail as well. (not sure why it can work in develop-upstream)

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc.bak (hunk 3 above)
Comment (jayfurmanek) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc.bak]: nope, will remove

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc (hunk 1 above)
Comment (jayfurmanek) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:78]: not sure I follow

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc [hunk 5]
```diff
@@ -101,7 +106,7 @@ xla::Status CompileAndPrintLlvmIr(const std::string& hlo_text,
                             llvm_module.get(), cuda_compute_capability,
                             hlo_module->config(), libdevice_dir));
     std::cout << ptx << std::endl;
-#else
+#elif TENSORFLOW_USE_ROCM
     std::string libdevice_dir = tsl::RocdlRoot();
     xla::gpu::GpuVersion gpu_version{rocm_compute_capability};
     TF_ASSIGN_OR_RETURN(
```
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:112]: line-113, ptx is nvidia thing, I think maybe we can change to std::vector<uint8_t> hasco_

Comment (jayfurmanek) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:112]: If you want, you can follow this PR up with one that cleans up the language and variable names in this file.

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc (hunk 4 above)
Comment (i-chaochen) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:95]: I think maybe we can move the initialization of platform_name and platform_id to there https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc#L74-L81

like this?

```
#if GOOGLE_CUDA
  std::string target_triple = "nvptx64-nvidia-cuda";
  std::string datalayout = "nvptx64-nvidia-cuda";
  std::string platform_name = "CUDA";
  stream_executor::Platform::Id platform_id =
      stream_executor::cuda::kCudaPlatformId;  
#else
  std::string target_triple = "amdgcn--amdhsa-amdgiz";
  std::string datalayout = "";
  std::string platform_name = "ROCm";
  stream_executor::Platform::Id platform_id =
      stream_executor::rocm::kROCmPlatformId;
#endif
  TF_ASSIGN_OR_RETURN(std::unique_ptr<llvm::Module> llvm_module,
                      xla::gpu::CompileModuleToLlvmIr(
                          hlo_module.get(), &llvm_context,
#if GOOGLE_CUDA
                          /*target_triple=*/xla::gpu::nvptx::TargetTriple(),
                          /*data_layout=*/xla::gpu::nvptx::DataLayout(),
#else
                          /*target_triple=*/xla::gpu::amdgpu::TargetTriple(),
                          /*data_layout=*/xla::gpu::amdgpu::DataLayout(),
#endif
                          /*platform_name=*/platform_name,
                          /*platform_id=*/platform_id, gpu_device_info,
                          cuda_compute_capability, rocm_compute_capability,
                          /*pointer_size=*/8));
```

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc (hunk 5 above)
Comment (i-chaochen) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:112]: Yes, I will keep following up.

File: tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc (hunk 4 above)
Comment (jayfurmanek) [reply tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:95]: Let's do this and your suggestion below about the variable names in a PR to upstream, with their final formatting stamp of approval then we'll sync it back here.

**Review by i-chaochen — APPROVED** [review]
